libraries/HumanPoseFuser.js

```diff
@@ -0,0 +1,803 @@
+            // make group from all poses with close proximity to this one
+            let ids = [];
+            for (let j of unassignedStandardIndices) {
+                if (proximityMatrix[index][j]) {
+                    ids.push(poseDataArr[j][0]);
+                }
+            }
+            poseObjsSamePerson.push(ids);
+        }
+
+        // simple approach: incremental merging groups of spatially-close poses into new fused human objects (one group at a time)
+        // WARNING: it can create multiple fused human objects for the same person when many cameras/apps are involved
+        // also, there can be issues with assigning to the same fused object a spatially stretched-out chain of poses
+        // TODO (future): proper proximity clustering with cycle consistency check
+        for (let ids of poseObjsSamePerson) {
+
+            if (ids.length < 2) {
+                // single standalone pose which is not close to any other, nothing to do in terms of fusion
+                continue;
+            }
+
+            // check if some objects in the group already belong to some fused human objects
+            let fusedObjectId = null;
+            let unassignedIds = [];
+            for (let id of ids) {
+                let fid = this.humanObjectsOfFusedObject.getFusedObject(id);
+                if (fid) {
+                    // NOTE: there could be human objects in this group which are already assigned to different fused human objects
+                    if (!fusedObjectId) {
+                        fusedObjectId = fid;
+                    }
+                } else {
+                    unassignedIds.push(id);
+                }
+            }
+
+            if (fusedObjectId) {
+                // add unassigned human objects to a fused human object at the same location
+                for (let id of unassignedIds) {
+                    this.humanObjectsOfFusedObject[fusedObjectId].push(id);
+
+                    // set parent reference pointing at a fused human object
+                    this.objectsRef[id].parent = fusedObjectId;
+
+                    this.batchedUpdates[id] = {
+                        objectKey: id,
+                        frameKey: null,
+                        nodeKey: null,
+                        propertyPath: 'parent',
+                        newValue: fusedObjectId,
+                        editorId: 0    // TODO: some server identificator
+                    };
+                    if (this.verbose) {
+                        console.log('assigning human obj=' + id + ' from ' + fusedObjectId);
+                    }
+                }
+            } else {
+                // whole group is unassigned, thus create new fused human object for it
+                // TODO?: add a string identifying the server  (equivalent to globalStates.tempUuid for devices)
+                let fusedObjectId = '_HUMAN_' + 'server' + '_pose1' + utilities.uuidTime();
```

Comment:
See TODO comment. Any suggestions about identificator of these server-based human objects? Do we actually need some server id unless we expect multiple servers in play for the same world object?

Reply:
For the time being this feels reasonable enough. Existing unique things about the server are getIP and the edge agent network id, of which getIP is the easiest to use.

The main case I'm worried about (and can't test) is 2 phone servers both doing human pose fusing and both fused human pose objects showing up in the remote operator.

The most durable approach for the future would be to only fuse human pose objects if the world object they're positioned relative to is hosted on the current server. This approach could then use the world id as part of its id (and keep track of how many fused pose objects there are to differentiate pose1/pose2/etc.)

Reply:
I didn't know, that the system would currently allow to have the server on phone doing the human pose fusion. So the other phones would stream poses to some 'master' device.

The number in 'pose1' is currently redundant since we distinguish the human objects by 'hash' from creation time. Actually, it is bit odd that the object id building is split for the standard human objects made by the app. In UI [code](https://github.com/ptcrealitylab/vuforia-spatial-toolbox-userinterface/blob/5d6949ca0e4de41b3a308d8173ebd2648637f265/src/humanPose/index.js#L33), it creates '_HUMAN_device3zdi63z1_pose1'  and then the server [code](https://github.com/ptcrealitylab/vuforia-spatial-edge-server/blob/cab85be7478770a6bd388b5e43317dfd255840c2/server.js#L2490) expands it to '_HUMAN_device3zdi63z1_pose1_Lgk3swbv6nx'. Maybe we should move it all to UI on app side.


Once we start really supporting multiple people, then there will be several distinct poses at the same timestamp and we need the counter to distinguish them. This applies to the human objects created by the app and fused human objects created on server.

Reply:
@hobinjk-ptc This just made me realize – I'm not sure if the current apps are discovering other worlds that are hosted on the local iOS server of another phone. I've only been testing with multiple devices that connect to a server on an external laptop.